### PR TITLE
FED-756 Allow analyzer 2.x, fix analyzer plugin in newer SDKs

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
## Motivation
We want to pull in newer analyzer versions, and unlocking analyzer 2.x is the next step.

Also, while testing I noticed that the analyzer plugin doesn't correctly start in IDEs in Dart 2.18 ☹️. Since this PR touches analyzer_plugin, I figured I'd fix that so we can manually verify changes.

## Changes
- Widen analyzer range to allow 2.x
- Reinstate validate_analyzer CI check to ensure over_react is compatible with both 1.x and 2.x
- Mockito
    - Pin mockito (in dev_dependencies) to work around compilation issue
    - Remove unnecessary List mock, which was causing build failures (more info in 578132d9eaf0f1df8547937f3c5bc25a103e0926)
- Upgrade analyzer plugin to analyzer 2.x. 
    Supporting both isn't necessary since the plugin is its own separate dependency graph, and supporting one is almost definitely easier. I went ahead and picked 2.x as the one to resolve to, since we'd have to upgrade anyways whenever the main package drops support for 1.x.
    - Fix warnings in caused by upgrade (mostly removal of non-null assertions, since analyzer 2.x promoted some APIs to non-nullable)
    - Copy IgnoreInfo into package, since we were previously importing it from analyzer via a src import, and the API changed to take in `ErrorCode` objects instead of Strings, and we couldn't construct ErrorCodes without more src imports
    - Fix low-hanging deprecation warnings
    - Disable ToggleComponentStatefulness, BoilerplateValidatorDiagnostic, and IncorrectDocCommentLocationDiagnostic ☹️. 
        These all import non-null-safe code from over_react's boilerplate parsing, which apparently prevents the plugin from being able to run in 2.18 SDK, and our previous workaround of a language comment in the plugin's `main` no longer works. Until we're able to migrate that code to null safety (which will hopefully be soonish), we'll need to stop importing these to get the plugin working again.

#### Release Notes
Widen analyzer range to allow 2.x in addition to 1.x

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        - All CI passes , notably the _Dart CI / validate_analyzer (2.18.7, ^2.0.0)_ run
        - Verify analyzer plugin works in playground in Dart 2.18
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Frameworks Design member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: tools/analyzer_plugin/lib/src/async_plugin_apis/diagnostic.dart [hunk 1]
```diff
@@ -68,21 +69,15 @@ mixin DiagnosticMixin on ServerPlugin {
     final analysisOptions = pluginOptionsReader.getOptionsForResult(analysisResult);
 
     try {
-      // If there is no relevant analysis result, notify the analyzer of no errors.
-      if (analysisResult.unit == null) {
-        channel.sendNotification(plugin.AnalysisErrorsParams(analysisResult.path!, []).toNotification());
-        return [];
-      }
-
```
Comment (robbecker-wf) [inline tools/analyzer_plugin/lib/src/async_plugin_apis/diagnostic.dart:76]: I assume this was removed because analysisResult.unit is non-nullable?

**Review by robbecker-wf — APPROVED** [review]

File: tools/analyzer_plugin/lib/src/async_plugin_apis/diagnostic.dart (hunk 1 above)
Comment (greglittlefield-wf) [reply tools/analyzer_plugin/lib/src/async_plugin_apis/diagnostic.dart:76]: Yup, exactly.

**Review by hunterbreathat-wk — APPROVED** [review]
While I may not be overly familiar with analyzers, these changes make sense; many of them are null-safe changes. I've got a couple of easy questions and a nit.

File: tools/analyzer_plugin/lib/src/util/ignore_info.dart [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        if (beforeMatch.trim().isEmpty) {
+          // The comment is on its own line, so it refers to the next line.
+          lineNumber++;
+        }
+        _ignoredOnLine
+            .putIfAbsent(lineNumber, () => [])
+            .addAll(comment.ignoredElements);
+      } else if (lexeme.contains('ignore_for_file:')) {
+        _ignoredForFile.addAll(comment.ignoredElements);
+      }
```
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply tools/analyzer_plugin/lib/src/util/ignore_info.dart:124]: This file was just copied from the analyzer package, so I left most of the logic as-is

File: .github/workflows/dart_ci.yml [hunk 3]
```diff
@@ -82,6 +82,50 @@ jobs:
...
+        # Can't run on `dev` (Dart 3) until we're fully null-safe.
+        # We can't run on Dart >=2.19 until https://github.com/dart-lang/sdk/issues/51128
+        # is fixed, or we work around it, which requires upgrading to analyzer ^5.1.0.
+        sdk: [ 2.13.4, 2.18.7 ]
+        analyzer:
+          - ^1.0.0
+          - ^2.0.0
+        # For 2.13.4, only validate analyzer ^1.0.0, since ^2.0.0 can't resolve.
+        exclude:
+          - sdk: 2.13.4
```
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply .github/workflows/dart_ci.yml:99]: Good point; the intention is to exclude the 2.13.4/^2.0.0 combination by excluding 2.13.4 and re-including this, but I think I could just exclude that combination directly. I'll take a stab at simplifying this

File: tools/analyzer_plugin/lib/src/diagnostic/analyzer_debug_helper.dart [hunk 4]
```diff
@@ -24,7 +24,7 @@ class AnalyzerDebugHelper {
   ///     debug.log('message');
   void log(String message) {
     if (!enabled) return;
-    collector.addError(code, Location(result.path!, 0, 0, 1, 1, 1, 1), errorMessageArgs: [message]);
```
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply tools/analyzer_plugin/lib/src/diagnostic/analyzer_debug_helper.dart:27]: The location removed the last two positional parameters (endLine and endColumn) in a breaking change and made them optional, so this just removes them.

This location is a bit bogus, and is for debugging purposes; it just adds a hint to the first character in the file.

[the remark this answers is not in the document]
Comment (hunterbreathat-wk) [reply tools/analyzer_plugin/lib/src/diagnostic/analyzer_debug_helper.dart:27]: Thanks!

File: .github/workflows/dart_ci.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply .github/workflows/dart_ci.yml:99]: Addressed in 640c5f57b987e5055f350a76b4c7a86d80de6c43

**Review by hunterbreathat-wk — APPROVED** [review]

**Comment by robbecker-wf** [comment]
QA+1 CI Passes with updated tests
We walked through using the analyzer plugin in a project and verified that it is using analyzer 2 and functions as it should.

**Comment by greglittlefield-wf** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**